cloud/azure/app-gateway/detectors-app-gateway.tf

```diff
@@ -0,0 +1,258 @@
+	}
+}
+
+resource "signalfx_detector" "http_5xx_errors" {
+	name = "${join("", formatlist("[%s]", var.prefixes))}[${var.environment}] Azure App Gateway 5xx error rate"
+
+	program_text = <<-EOF
+		from signalfx.detectors.aperiodic import conditions
+		A = data('ResponseStatus', filter=filter('resource_type', 'Microsoft.Network/applicationGateways') and filter('httpstatusgroup', '5xx') and filter('primary_aggregation_type', 'true') and ${module.filter-tags.filter_custom})${var.http_5xx_errors_aggregation_function}
+		B = data('ResponseStatus', filter=filter('resource_type', 'Microsoft.Network/applicationGateways') and filter('primary_aggregation_type', 'true') and ${module.filter-tags.filter_custom})${var.http_5xx_errors_aggregation_function}
+		signal = ((A/B)*100).${var.http_5xx_errors_transformation_function}(over='${var.http_5xx_errors_transformation_window}').publish('signal')
+		ON_Condition_CRIT = conditions.generic_condition(signal, ${var.http_5xx_errors_threshold_critical}, ${var.http_5xx_errors_threshold_critical}, 'above', lasting('${var.http_5xx_errors_aperiodic_duration}', ${var.http_5xx_errors_aperiodic_percentage}), 'observed')
+		ON_Condition_WARN = conditions.generic_condition(signal, ${var.http_5xx_errors_threshold_warning}, ${var.http_5xx_errors_threshold_critical}, 'within_range', lasting('${var.http_5xx_errors_aperiodic_duration}', ${var.http_5xx_errors_aperiodic_percentage}), 'observed', strict_2=False)
+		detect(ON_Condition_CRIT, off=when(signal is None, '${var.http_5xx_errors_clear_duration}')).publish('CRIT')
+		detect(ON_Condition_WARN, off=when(signal is None, '${var.http_5xx_errors_clear_duration}')).publish('WARN')
+	EOF
+
+	rule {
+		description           = "is too high > ${var.http_5xx_errors_threshold_critical}"
+		severity              = "Critical"
+		detect_label          = "CRIT"
+		disabled              = coalesce(var.http_5xx_errors_disabled_critical, var.http_5xx_errors_disabled, var.detectors_disabled)
+		notifications         = coalescelist(var.http_5xx_errors_notifications_critical, var.http_5xx_errors_notifications, var.notifications)
+		parameterized_subject = "[{{ruleSeverity}}]{{{detectorName}}} {{{readableRule}}} ({{inputs.signal.value}}) on {{{dimensions}}}"
+	}
+
+	rule {
+		description           = "is too high > ${var.http_5xx_errors_threshold_warning}"
+		severity              = "Warning"
+		detect_label          = "WARN"
+		disabled              = coalesce(var.http_5xx_errors_disabled_warning, var.http_5xx_errors_disabled, var.detectors_disabled)
+		notifications         = coalescelist(var.http_5xx_errors_notifications_warning, var.http_5xx_errors_notifications, var.notifications)
+		parameterized_subject = "[{{ruleSeverity}}]{{{detectorName}}} {{{readableRule}}} ({{inputs.signal.value}}) on {{{dimensions}}}"
+	}
+}
+
+resource "signalfx_detector" "backend_http_4xx_errors" {
```

Comment:
Still not working as expected.

I've tried in UI with this given formula that seems to work fine:
```
from signalfx.detectors.aperiodic import conditions
A = data('BackendResponseStatus', extrapolation="zero", filter=filter('resource_type', 'Microsoft.Network/applicationGateways') and filter('httpstatusgroup', '4xx') and filter('primary_aggregation_type', 'true') and filter('azure_tag_env', 'preprod') and filter('azure_tag_sfx_monitored', 'true')).sum(by=['azure_resource_name', 'azure_resource_group_name', 'azure_region', 'backendhttpsetting', 'backendpool', 'backendserver'])
B = data('BackendResponseStatus', filter=filter('resource_type', 'Microsoft.Network/applicationGateways') and filter('primary_aggregation_type', 'true') and filter('azure_tag_env', 'preprod') and filter('azure_tag_sfx_monitored', 'true')).sum(by=['azure_resource_name', 'azure_resource_group_name', 'azure_region', 'backendhttpsetting', 'backendpool', 'backendserver'])
signal = ((A/B)*100).fill(0).min(over='5m').publish('signal')
detect(when(signal > 95)).publish('CRIT')
detect(when(signal > 80) and when(signal <= 95)).publish('WARN')
```

I extrapolate the 4xx errors with zero values (no 4xx metric equal to 0 4xx requests) and also fill with 0 the operation in order to have a 0 value when having a divide by zero operation in signalflow.

Does this look fine to you ? If so, can you use this pattern on other 4xx and 5xx requests detector for application gateway. Also, it seems accurate for other Azure components.

Reply:
Also, we need a sum rollup on the data.